python/gps/algorithm/algorithm_traj_opt.py

```diff
@@ -169,38 +165,33 @@ def _update_cost(self):
             else:
                 i = idx
             # Evaluate demo prob. under sample distributions.
-            for itr_i in xrange(itr):
+            for itr_i in xrange(itr + 1):
                 traj = self.traj_distr[itr_i][i]
                 for j in xrange(demoX[idx].shape[0]):
                     for t in xrange(self.T - 1):
                         diff = traj.k[t, :] + \
                                 traj.K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
-                        demos_prob[idx][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                        demos_logprob[idx][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
                                                         np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
             # Evaluate demo prob. under demo distributions.
             for itr_i in xrange(Md):
                 for j in range(demoX[idx].shape[0]):
                     for t in xrange(self.T - 1):
                         diff = demo_traj[itr_i].k[t, :] + \
                                 demo_traj[itr_i].K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
-                        demos_prob[idx][itr + itr_i, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                        demos_logprob[idx][itr + itr_i + 1, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
                                                         np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
-            # Sum over time.
-            demos_prob_T[idx] = np.sum(demos_prob[idx], 1)
-            # Sum over the distributions.
-            demos_sum[idx] = logsum(demos_prob_T[idx], 0)
-
-        self.samples_prob = samples_sum
-        self.demos_prob = demos_sum # The matlab code doesn't store this, which is confusing.
+            # Sum over the distributions and time.
+            demos_logiw[idx] = logsum(np.sum(demos_logprob[idx], 1), 0)
 
-        # For testing purposes.
-        print "sample sum iteration " + "itr: " + samples_sum
-        print "demo sum iteration " + "itr: " + demos_sum
 
         # Update the learned cost
+        sampleU = {i: self.cur[i].sample_list.get_U() for i in xrange(M)}
+        sampleX = {i: self.cur[i].sample_list.get_X() for i in xrange(M)}
+        sampleO = {i: self.cur[i].sample_list.get_obs() for i in xrange(M)}
         cost_ioc_quad = self.cost # set the type of cost to be cost_ioc_quad here.
-        self.learned_cost = cost_ioc_quad.update(demoU, demoX, self.demoO, demos_sum, sampleU, sampleX,\
-                                                sampleO, samples_sum, itr, eta, self) # still need to get sample U/X/O and eta
+        self.learned_cost = cost_ioc_quad.update(demoU, demoX, demoO, demos_logiw, sampleU, sampleX,\
```

Comment:
We learn a single cost (cost_ioc_quad) across all conditions. So, we just need a line of code for each object that concatenates the conditions together.

It's possible that, in the future, we may want to try learning different costs per condition, but that would be easy to add later.